edx/edx-enterprise
# Disable atomic transactions for CourseEnrollmentView

This ensures that the EnterpriseCustomerUser created during the `get()` response is
available to the E-Commerce API calls when fetching the final prices.

This issue causes the prices that appear on the Enterprise Enrollment page to appear without taking into account any entitlement discounts which apply for the Enterprise Customer when the page is first displayed, because the E-Commerce API doesn't see the `EnterpriseCustomerUser` entry in the database until after the page is rendered.  Refreshing the Enterprise Enrollment page rectifies the issue, but is not a good user experience.

**JIRA:** [ENT-460](https://openedx.atlassian.net/browse/ENT-460)

**Dependencies:** None

**Merge deadline:** 04/Jul/17

**Installation instructions:** 

Requires a stack with ecommerce and course-discovery services configured.

To enable enterprise on ecommerce, you'll need

1. Create an Enterprise
1. Create a course
1. Add the course to ecommerce, ensuring it has a premium mode
1. Create an Enterprise-linked coupon in ecommerce, valid for the previously-create course.  
   Ensure that the coupon has Enrollment Type: Discount Code.
1. Add the coupon as an entitlement to the EnterpriseCustomer in the LMS admin

**Testing instructions:**

1. Create a new user, unlinked to the EnterpriseCustomer.
   The new user can register via TPA/SSO, or through the standard registration process.
1. Using the new user, navigate to the Enterprise course enrollment page, e.g. 
   `/enterprise/b38ccece-c5c4-4bbb-b6f2-c32b6c9cfd50/course/course-v1%3AOpenCraft%2BENT%2B460/enroll/`
1. Note that the price for the premium mode always appears with the coupon applied.

Be careful if you intend to re-test with the same user, since this issue disappears after the first enrollment page view:

1. Use the LMS Admin - EnterpriseCustomer's Manage Learners view to un-link the user from the Enterprise Customer before refreshing the Enterprise Enrollment view page.
1. Beware of the ecommerce `ENTERPRISE_API_CACHE_TIMEOUT` setting; it will preserve the Enterprise Learner Data fetched from the enterprise API for that many seconds.

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).  cf 3c2a0e1
- [x] Documentation updated (not only docstrings): bug fix only, no docs changes required.
- [x] <strike>Translations are updated</strike> No user-facing strings added or changed.

**Post merge:**

**Author concerns:**

This `@transaction.non_atomic_requests` mechanism isn't pretty, but I've tried to explain its necessity and usage in the comments added.  It is also used in several key places in edx-platform, notably [`course_mode.views.ChooseModeView`](https://github.com/edx/edx-platform/blob/53cd05fb7fcb3f45d4d8868945c6912e95efd3ca/common/djangoapps/course_modes/views.py#L46).

[`ChooseModeView.get()` also takes advantage](https://github.com/edx/edx-platform/blob/53cd05fb7fcb3f45d4d8868945c6912e95efd3ca/common/djangoapps/course_modes/views.py#L214) of edx-platform's [`db.utils.outer_atomic` decorator](https://github.com/edx/edx-platform/blob/e8a36957b1f732974260e7b9b42b9c25148b492c/common/djangoapps/util/db.py#L220), which will throw if it's nested inside a `transaction.atomic` block, thus protecting the atomicity of the transaction even further.  However, I decided against using `db.utils.outer_atomic` in this case, because it creates another dependency on the edx-platform repo, and it's an almighty pain to stub out decorators.
